File: entsoe-client/src/main/java/org/example/model/load/GLMarketDocument.java
package org.example.model.load;

import java.time.LocalDateTime;
import javax.xml.bind.annotation.XmlAccessType;
import javax.xml.bind.annotation.XmlAccessorType;
import javax.xml.bind.annotation.XmlElement;
import javax.xml.bind.annotation.XmlRootElement;
import javax.xml.bind.annotation.adapters.XmlJavaTypeAdapter;
import lombok.Data;
import lombok.NoArgsConstructor;
import org.example.adapter.*;
import org.example.model.common.*;

// Root element for Load Domain responses
@XmlRootElement(
    name = "GL_MarketDocument",
    namespace = "urn:iec62325.351:tc57wg16:451-6:generationloaddocument:3:0")
@XmlAccessorType(XmlAccessType.FIELD)
@Data
@NoArgsConstructor
public class GLMarketDocument {
  static final String XML_NAMESPACE = "urn:iec62325.351:tc57wg16:451-6:generationloaddocument:3:0";

  @XmlElement(name = "mRID", namespace = XML_NAMESPACE)
  private String mRID;

  @XmlElement(name = "revisionNumber", namespace = XML_NAMESPACE)
  private Integer revisionNumber;

  @XmlElement(name = "type", namespace = XML_NAMESPACE)
  @XmlJavaTypeAdapter(DocumentTypeAdapter.class)
  private DocumentType type;

  @XmlElement(name = "process.processType", namespace = XML_NAMESPACE)
  @XmlJavaTypeAdapter(ProcessTypeAdapter.class)
  private ProcessType processType;

  @XmlElement(name = "sender_MarketParticipant.mRID", namespace = XML_NAMESPACE)
  private MarketParticipantMRID senderMarketParticipantMRID;

  @XmlElement(name = "sender_MarketParticipant.marketRole.type", namespace = XML_NAMESPACE)
  @XmlJavaTypeAdapter(MarketRoleTypeAdapter.class)
  private MarketRoleType senderMarketParticipantMarketRoleType;

  @XmlElement(name = "receiver_MarketParticipant.mRID", namespace = XML_NAMESPACE)
  private MarketParticipantMRID receiverMarketParticipantMRID;

  @XmlElement(name = "receiver_MarketParticipant.marketRole.type", namespace = XML_NAMESPACE)
  @XmlJavaTypeAdapter(MarketRoleTypeAdapter.class)
  private MarketRoleType receiverMarketParticipantMarketRoleType;

  @XmlElement(name = "createdDateTime", namespace = XML_NAMESPACE)
  @XmlJavaTypeAdapter(LocalDateTimeAdapter.class)
  private LocalDateTime createdDateTime;

  @XmlElement(name = "time_Period.timeInterval", namespace = XML_NAMESPACE)
  private LoadTimeInterval timePeriodTimeInterval;

  @XmlElement(name = "TimeSeries", namespace = XML_NAMESPACE)
  private LoadTimeSeries timeSeries;

  // Helper method to get created date as LocalDateTime
  //    public LocalDateTime getCreatedDateTimeAsLocalDateTime() {
  //        if (createdDateTime != null) {
  //            return LocalDateTime.parse(createdDateTime, DateTimeFormatter.ISO_DATE_TIME);
  //        }
  //        return null;
  //    }
}

File: entsoe_client/src/entsoe_client/model/load/gl_market_document.py
from datetime import datetime

from pydantic import field_serializer, field_validator
from pydantic_xml import BaseXmlModel, element

from entsoe_client.adapters import date_time_adapter
from entsoe_client.model import ENTSOE_LOAD_NSMAP
from entsoe_client.model.common.document_type import DocumentType
from entsoe_client.model.common.market_role_type import MarketRoleType
from entsoe_client.model.common.process_type import ProcessType
from entsoe_client.model.load.load_time_interval import LoadTimeInterval
from entsoe_client.model.load.load_time_series import LoadTimeSeries
from entsoe_client.model.load.market_participant_mrid import MarketParticipantMRID


class GlMarketDocument(BaseXmlModel, tag="GL_MarketDocument", nsmap=ENTSOE_LOAD_NSMAP):  # type: ignore[call-arg]
    mRID: str = element(tag="mRID")
    revisionNumber: int | None = element(tag="revisionNumber")
    type: DocumentType = element(tag="type")
    processType: ProcessType = element(tag="process.processType")
    senderMarketParticipantMRID: MarketParticipantMRID = element(
        tag="sender_MarketParticipant.mRID",
    )
    senderMarketParticipantMarketRoleType: MarketRoleType = element(
        tag="sender_MarketParticipant.marketRole.type",
    )
    receiverMarketParticipantMRID: MarketParticipantMRID = element(
        tag="receiver_MarketParticipant.mRID",
    )
    receiverMarketParticipantMarketRoleType: MarketRoleType = element(
        tag="receiver_MarketParticipant.marketRole.type",
    )
    createdDateTime: datetime = element(tag="createdDateTime")
    timePeriodTimeInterval: LoadTimeInterval = element(tag="time_Period.timeInterval")
    timeSeries: list[LoadTimeSeries]

    @field_serializer("type")  # type: ignore[misc]
    def encode_type(self, value: DocumentType) -> str:
        return value.code

    @field_validator("type", mode="before")  # type: ignore[misc]
    def decode_type(
        cls,
        value: str | DocumentType | None,
    ) -> DocumentType | None:
        if isinstance(value, str):
            return DocumentType.from_code(value)

        return value

    @field_serializer("processType")  # type: ignore[misc]
    def encode_process_type(self, value: ProcessType) -> str:
        return value.code

    @field_validator("processType", mode="before")  # type: ignore[misc]
    def decode_process_type(
        cls,
        value: str | ProcessType | None,
    ) -> ProcessType | None:
        if isinstance(value, str):
            return ProcessType.from_code(value)

        return value

    @field_serializer("senderMarketParticipantMarketRoleType")  # type: ignore[misc]
    def encode_sender_role_type(self, value: MarketRoleType) -> str:
        return value.code

    @field_validator("senderMarketParticipantMarketRoleType", mode="before")  # type: ignore[misc]
    def decode_sender_role_type(
        cls,
        value: str | MarketRoleType | None,
    ) -> MarketRoleType | None:
        if isinstance(value, str):
            return MarketRoleType.from_code(value)

        return value

    @field_serializer("receiverMarketParticipantMarketRoleType")  # type: ignore[misc]
    def encode_receiver_role_type(self, value: MarketRoleType) -> str:
        return value.code

    @field_validator("receiverMarketParticipantMarketRoleType", mode="before")  # type: ignore[misc]
    def decode_receiver_role_type(
        cls,
        value: str | MarketRoleType | None,
    ) -> MarketRoleType | None:
        if isinstance(value, str):
            return MarketRoleType.from_code(value)

        return value

    @field_serializer("createdDateTime")  # type: ignore[misc]
    def encode_created_datetime(self, value: datetime) -> str:
        return date_time_adapter.encode_content(value)

    @field_validator("createdDateTime", mode="before")  # type: ignore[misc]
    def decode_created_datetime(
        cls,
        value: str | datetime | None,
    ) -> datetime | None:
        if isinstance(value, str):
            return date_time_adapter.decode_content(value)

        return value
